Documentation: Updated Docker installation instructions

co-decode opened the pull request:
## Checklist

#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://intranet.neuro.polymtl.ca/geek-tips/contributing#pr-labels-a-href-pr-labels-id-pr-labels-a) to this PR
- [ ] I've applied a [release milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [ ] I've assigned a reviewer

#### PR contents

- [x] I've consulted [SCT's internal developer documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Testing) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
Issue #4101 mentions that the Docker-related installation instructions are out of date.

I have: 
- Removed CentOS instructions as mentioned.
- Tested and updated instructions for Installation with the Ubuntu 22.04 LTS docker image.

**Linux**
- Tested and updated instructions for Forwarding the Xorg X11 server on Linux.

**Windows**
- Tested and updated instructions for installing and using the Xming X11 server on Windows.
- Removed instructions for Docker Toolbox, as Toolbox is now deprecated.
- Tested and removed instructions pertaining to the Xming/ssh Docker Desktop XLaunch script; I was unable to get the script working.

**Mac**
- Updated instructions, but unfortunately I was unable to test them on MacOS.

### Comments
PyQt requires additional dependencies now. If even more dependencies are needed in the future, the output of `sct_check_dependencies` run with the environment variable `QT_DEBUG_PLUGINS=1` set will help to diagnose what is missing. 

## Linked issues

Resolves #4101

joshuacwnewton commented:
Thank you so much for taking a look at this issue! I'll be happy to take a look whenever this PR is ready for review. :relaxed:

joshuacwnewton commented on documentation/source/user_section/installation/linux.rst:
```diff
@@ -214,14 +192,32 @@ First, save your Docker image:
-4. In XQuartz window xhost + 127.0.0.1
-5. In your other Terminal window, run:
-   ``docker run -ti --rm -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix <CONTAINER_ID>``
+3. In your Terminal window, run:
+   ``docker run -it --rm --privileged -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix <CONTAINER_ID>``
```
Hmm. After I got the `docker run` command working, I tried running `sct_check_dependencies` to see if PyQt would work. But, I got the following error:

```
Check if figure can be opened with PyQt.............qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, webgl, xcb.
```

I then tried again with `export QT_DEBUG_PLUGINS=1` specified (as you suggested), and got the following output:

<details>

```
Check if figure can be opened with PyQt.............QFactoryLoader::QFactoryLoader() checking directory path "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms" ...
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqeglfs.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqeglfs.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "eglfs"
        ]
    },
    "archreq": 0,
    "className": "QEglFSIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("eglfs")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqlinuxfb.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqlinuxfb.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "linuxfb"
        ]
    },
    "archreq": 0,
    "className": "QLinuxFbIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("linuxfb")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqminimal.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqminimal.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "minimal"
        ]
    },
    "archreq": 0,
    "className": "QMinimalIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("minimal")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqminimalegl.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqminimalegl.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "minimalegl"
        ]
    },
    "archreq": 0,
    "className": "QMinimalEglIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("minimalegl")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqoffscreen.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqoffscreen.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "offscreen"
        ]
    },
    "archreq": 0,
    "className": "QOffscreenIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("offscreen")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqvnc.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqvnc.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "vnc"
        ]
    },
    "archreq": 0,
    "className": "QVncIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("vnc")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwayland-egl.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwayland-egl.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "wayland-egl"
        ]
    },
    "archreq": 0,
    "className": "QWaylandEglPlatformIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("wayland-egl")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwayland-generic.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwayland-generic.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "wayland"
        ]
    },
    "archreq": 0,
    "className": "QWaylandIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("wayland")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwayland-xcomposite-egl.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwayland-xcomposite-egl.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "wayland-xcomposite-egl"
        ]
    },
    "archreq": 0,
    "className": "QWaylandXCompositeEglPlatformIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("wayland-xcomposite-egl")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwayland-xcomposite-glx.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwayland-xcomposite-glx.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "wayland-xcomposite-glx"
        ]
    },
    "archreq": 0,
    "className": "QWaylandXCompositeGlxPlatformIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("wayland-xcomposite-glx")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwebgl.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqwebgl.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "webgl"
        ]
    },
    "archreq": 0,
    "className": "QWebGLIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("webgl")
QFactoryLoader::QFactoryLoader() looking at "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqxcb.so"
Found metadata in lib /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqxcb.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "xcb"
        ]
    },
    "archreq": 0,
    "className": "QXcbIntegrationPlugin",
    "debug": false,
    "version": 330752
}


Got keys from plugin meta data ("xcb")
QFactoryLoader::QFactoryLoader() checking directory path "/sct/python/envs/venv_sct/bin/platforms" ...
Cannot load library /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqxcb.so: (libXrender.so.1: cannot open shared object file: No such file or directory)
QLibraryPrivate::loadPlugin failed on "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqxcb.so" : "Cannot load library /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqxcb.so: (libXrender.so.1: cannot open shared object file: No such file or directory)"
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, webgl, xcb.
```
</details>

The key line is:

```
Cannot load library /sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqxcb.so: (libXrender.so.1: cannot open shared object file: No such file or directory)
```

To fix this, I ran the following command inside the container:

```
apt install libxrender1
```

I then re-ran `sct_check_dependencies` but got the following new error:

```
QFactoryLoader::QFactoryLoader() checking directory path "/sct/python/envs/venv_sct/bin/platforms" ...
loaded library "/sct/python/envs/venv_sct/lib/python3.9/site-packages/PyQt5/Qt/plugins/platforms/libqxcb.so"
qt.qpa.xcb: could not connect to display :0.0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```

I am stuck here. I have tried the following solutions:

- Running [`xhost +local:docker`](https://stackoverflow.com/questions/28392949/running-chromium-inside-docker-gtk-cannot-open-display-0#comment57024354_28395350)
- Running [`xhost local:root`](https://stackoverflow.com/a/34586732)
- Adding [`:rw`](https://stackoverflow.com/q/69872788) to the `-v` option.
- Adding [`--net=host`](https://unix.stackexchange.com/a/707724/583446) to the `docker run` command.
- Installing [additional dependencies](https://stackoverflow.com/a/76610935).

But all of this results in the same output. If you have any ideas, please let me know. :slightly_smiling_face:

co-decode replied:
Thanks for your prompt and thorough reply @joshuacwnewton .
I have been testing on Arch Linux with the Docker CLI, and it seems as though there are differences between Docker Desktop and the Docker CLI.
- libxrender was not needed on my Arch machine, but I found that it was necessary for my Windows machine (Docker Desktop).
- the /tmp/.X11-unix directory was permitted as shared by default on my Arch machine, I found that I didn't need to mount the X11 Unix socket on Windows.

It seems like it would be something to do with the X11 forwarding through Docker Desktop. I am currently installing Docker Desktop on my Arch machine and will see if I can find anything.

joshuacwnewton replied:
Ahhh, thank you so much for the investigation, and for examining the various difference between OSs and versions of Docker. :hearts:

joshuacwnewton reviewed (changes requested):
Here is an initial review for the Linux instructions. (I tried it on my Debian 11 laptop.)

I will try to review Mac and Windows as well later on. Before I review, though, could you double check to see if any of my feedback from the Linux instructions also applies to the Mac/Windows instructions?

Thank you so much again! :heart:

co-decode commented:
I have implemented the changes suggested above and rectified the typo in the libxkbcommon-x11-0 dependency.

During testing on Debian, I was unable to get an X server forwarded to the container in Docker Desktop for Linux.
According to [this thread on the Docker Forums](https://forums.docker.com/t/is-it-possible-to-display-x11-through-socketing-in-docker-desktop-for-linux/133044), it seems like X forwarding through Docker Desktop for Linux might be unsupported, or else very complicated.

Fortunately, I was able to install Docker Community Edition Engine from the _apt_ repository and the X11 forwarding then worked through xWayland with no further issues.

I have recommended against the use of Docker Desktop for Linux in the documentation, and have pointed users to the Docker CE Engine.

joshuacwnewton commented on documentation/source/user_section/installation/linux.rst:
```diff
@@ -145,7 +145,14 @@ In the context of SCT, it can be used:
+   Instead install the `Docker CE Engine <https://docs.docker.com/engine/install>`_, which is separate to the Docker Desktop Engine.
+   For example on Ubuntu/Debian, follow the instructions for installing Docker from the `apt repository <https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository>`_.
+
+Then, follow the example below to create an OS-specific SCT installation.
```
I tried the instructions again from scratch with Docker Server, and realized that the commands in the example need `sudo`, unless the user [creates the `docker` group](https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user).

We might want to add another section (between "install docker" and "run the docker steps") that specifically mentions the `sudo` workaround?

```suggestion
By default, Docker commands require the use of ``sudo`` for additional permissions. If you want to run Docker commands without needing to add ``sudo`` (as shown in the example below), please follow `these instructions <https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user>`_ to create a Unix group called ``docker``, then add users your user account to it.

Then, you can follow the example below to create an OS-specific SCT installation.
```

joshuacwnewton commented on documentation/source/user_section/installation/linux.rst:
```diff
@@ -154,52 +161,30 @@ Docker Image: Ubuntu
    # Test SCT
    sct_testing
    # save the state of the container. Open a new Terminal and run:
    docker ps -a  # list all containers
-   docker commit <CONTAINER_ID> <YOUR_NAME>/centos:centos7
+   docker commit <CONTAINER_ID> <YOUR_NAME>/ubuntu:ubuntu22.04
```
This is kind of a nitpick, but `<YOUR_NAME>` might be a bit unclear? (If a user takes it extremely literally, they might put e.g. `joshua` instead of a useful image name :rofl:)

So, it might be better to explicitly state that the user is choosing the name for a docker image here:

```bash
   # save the state of the container as a docker image. 
   # Open a new terminal and run:
   docker ps -a  # list all containers (to find out the container ID)
   # specify the ID, and also choose a name to use for the docker image, such as "sct_v6.0"
   docker commit <CONTAINER_ID> <IMAGE_NAME>/ubuntu:ubuntu22.04
```

(I also updated the comments in this section to be a bit more explicit, too.)

((This would also require changing `<YOUR_NAME>` to `<IMAGE_NAME>` elsewhere in the documentation.))

joshuacwnewton commented on documentation/source/user_section/installation/linux.rst:
```diff
@@ -208,20 +193,38 @@ First, save your Docker image:
-4. In XQuartz window xhost + 127.0.0.1
-5. In your other Terminal window, run:
-   ``docker run -ti --rm -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix <CONTAINER_ID>``
+3. In your Terminal window, run:
+   ``docker run -it --rm --privileged -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix <YOUR_NAME>/ubuntu:ubuntu22.04``
```
Just for uniformity, it might be good to have this as a `..code ::` directive as well.

joshuacwnewton replied:
Also, in this section, it might be good to add an additional step 4 (for testing that GUI forwarding is working):

```suggestion
3. In your Terminal window, run:
   
   .. code:: docker run -it --rm --privileged -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix <YOUR_NAME>/ubuntu:ubuntu22.04``

4. Then, you can check if GUI scripts work by running the following command:

   .. code:: sct_check_dependencies

   If GUI scripts are enabled, you should see all green ``[OK]`` results without any errors.

joshuacwnewton reviewed (changes requested):
I had a bit more feedback for the `linux.rst` page (mostly small clarifications to working for clarity). But, everything worked well for me on my machine! :tada: 

I'll try to look at `windows.rst` next, and I'll also ask another lab member if they can try the `macos.rst` instructions as well.

co-decode commented:
- Docker CE Engine was changed to Docker Server Engine, and the links were updated. I've also mentioned that user's should install from their Linux distribution's repository.
- I've added `sudo` to the relevant host commands, and also kept the instructions for adding docker permissions to the non-root user.
- <YOUR_NAME> has been changed to <IMAGE_NAME>.
- The changes to the end of the X11 forwarding section have been included.

Thanks for the feedback, I appreciate the opportunity to contribute! I am an IT student with a background in the medical sciences, and this is an interesting project that helps me practice. I hope I can bring some value to the project.

joshuacwnewton commented:
> I am an IT student with a background in the medical sciences, and this is an interesting project that helps me practice. I hope I can bring some value to the project.

Your contributions have been very high quality, and are greatly appreciated!! I have worked with quite a few external collaborators over the years, and I think you have excellent git/GitHub + communication skills. You have been excellent to work with. :relaxed:

jcohenadad commented:
Thank you so much for your excellent contributions @co-decode 🙏  This is greatly appreciated!

joshuacwnewton commented on documentation/source/user_section/installation/mac.rst:
```diff
@@ -203,16 +183,19 @@ First, save your Docker image:
+   .. code:: bash 
+
+      docker run -e DISPLAY=host.docker.internal:0 -it <IMAGE_NAME>/ubuntu:ubuntu22.04
 
```
Adding a note here to add the same `sct_check_dependencies` step to the other pages:

```suggestion

6. You can test whether GUI scripts are available by running the following command in your Docker container:
 
   .. code:: bash
   
      sct_check_dependencies
      
   You should see two green ``[OK]`` symbols at the bottom of the report for "PyQT" and "matplotlib" checks, which represent the GUI features provided by SCT. 
   
   You may see some warnings about "current screen resources", "XDG_RUNTIME_DIR", or "FBConfigs". These are non-critical warnings, and should not affect the performance of the GUI scripts.
```

joshuacwnewton commented on documentation/source/user_section/installation/linux.rst:
```diff
@@ -154,74 +163,83 @@ Docker Image: Ubuntu
-4. In XQuartz window xhost + 127.0.0.1
-5. In your other Terminal window, run:
-   ``docker run -ti --rm -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix <CONTAINER_ID>``
+      sct_check_dependencies
+      
+   If GUI scripts are enabled, you should see all green ``[OK]`` results without any errors.
```
```suggestion

6. You can test whether GUI scripts are available by running the following command in your Docker container:
 
   .. code:: bash
   
      sct_check_dependencies
      
   You should see two green ``[OK]`` symbols at the bottom of the report for "PyQT" and "matplotlib" checks, which represent the GUI features provided by SCT. 
   
   You may see some warnings about "current screen resources", "XDG_RUNTIME_DIR", or "FBConfigs". These are non-critical warnings, and should not affect the performance of the GUI scripts.
```

joshuacwnewton reviewed (changes requested):
I have just tested the instructions on Windows, and everything works well for me. :tada:

My comments here are mainly just slight nitpicks / wording tweaks to make things clearer for our less-technical users. :slightly_smiling_face: 

I have also requested help from @mguaypaq to review the MacOS instructions, as I do not personally have access to a MacOS computer.

mguaypaq commented:
Just a quick comment with my current observations about testing on a Mac. I'll suggest changes next week (note that it's a long weekend here in Canada).
- I didn't have to test installing `gcc` because it was already installed on the Mac computer I have access to, apparently through Xcode?
  ```
  % gcc --version
  Apple clang version 14.0.0 (clang-1400.0.29.202)
  Target: x86_64-apple-darwin21.6.0
  Thread model: posix
  InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
  ```
- I also didn't have to install Docker: the `docker` command was possibly installed through Xcode already, and Docker Desktop was already installed manually. As expected, the `docker pull` command refused to work at first with a cryptic error message, because I wasn't running Docker Desktop at the same time:
  ```
  % docker pull ubuntu:22.04
  Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
  ```
  We should probably mention this exact error message and how to fix it in the instructions.
- After running the `./install_sct -y` step, the last thing on the screen is a couple of FAIL messages. Maybe they're expected? If so, we should reassure the users explicitly in the instructions:
  <img width="740" alt="image" src="https://github.com/spinalcordtoolbox/spinalcordtoolbox/assets/928742/18f9f961-2efd-4af0-a91b-2c56b734c67a">
- Running `sct_testing` seems to have worked, and that's as far as I got today:
  ```
  ========= 435 passed, 19 skipped in 473.66s (0:07:53) ========
  ```

joshuacwnewton commented:
> * After running the `./install_sct -y` step, the last thing on the screen is a couple of FAIL messages. Maybe they're expected? If so, we should reassure the users explicitly in the instructions:

I can confirm that these [FAIL] messages are expected. They will only change to [OK] after users run the optional "Add support for GUI scripts" steps. I agree that adding a note here would be helpful. :slightly_smiling_face:

co-decode commented on documentation/source/user_section/installation/windows.rst:
```diff
@@ -343,53 +326,22 @@ First, save your Docker image:
+   .. code:: bash 
+   
+      docker run -it --rm -e DISPLAY=<IP of vEthernet Adapter>:<Display Number> <IMAGE_NAME>/ubuntu:ubuntu22.04
 
```
I was getting the same output of non-fatal warnings when I ran sct_check_dependencies.
Fortunately I was able to install VcXsrv to launch the X11 server instead of Xming, and the warnings went away!
The public Xming server version was released in 2007, and the VcXsrv version I downloaded was from 2022.

I have:
- linked to the VcXsrv download and provided instructions to set it up correctly
- added an XDG_RUNTIME_DIR environment variable to the docker run command
This should make a clean <sct_check_dependencies> reproducible.

co-decode commented:
I've committed all the suggestions above, except for the mention of warnings related to "current screen resources", etc.
I removed the recommendation to download Xming, and now only recommend VcXsrv.
I've edited the instructions for starting an X11 server with VcXsrv, and I've added an environment variable to the `docker run` command.
Hopefully these changes can get rid of those warnings.

I have a question about the last instruction I included in the Docker instructions for Windows.
There I explained how to map ports between the host machine and the docker container, perhaps for the purposes of serving `.html` output from the container so that the host machine can access it the output through its web browser.

Is this a valid use case for SCT with Docker or would it be better to omit these instructions?

joshuacwnewton commented on documentation/source/user_section/installation/windows.rst:
```diff
@@ -343,53 +327,32 @@ First, save your Docker image:
+ 
+   .. code:: bash
+   
+      sct_check_dependencies
```
>  I removed the recommendation to download Xming, and now only recommend VcXsrv.

I tried the new instructions, and they mostly worked perfectly! I just ran into one final warning:

```
QStandardPaths: XDG_RUNTIME_DIR points to non-existing path '/tmp/runtime-root', please create it with 0700 permissions.
```

This fixed the issue:

```suggestion
      mkdir /tmp/runtime-root
      sct_check_dependencies
```

joshuacwnewton commented on documentation/source/user_section/installation/windows.rst:
```diff
@@ -343,53 +327,32 @@ First, save your Docker image:
+      
+   You should see two green ``[OK]`` symbols at the bottom of the report for "PyQT" and "matplotlib" checks, which represent the GUI features provided by SCT. 
+   
+If you plan to use a web server within the container to access SCT output from the host's browser, be sure to map your ports by including the ``-p <Host Port>:<Container Port>`` flag in the above command.
```
> There I explained how to map ports between the host machine and the docker container, perhaps for the purposes of serving `.html` output from the container so that the host machine can access it the output through its web browser.
> 
> Is this a valid use case for SCT with Docker or would it be better to omit these instructions?

SCT has a "QC Report" feature that uses the web browser. However, it doesn't run a web server -- instead, produces an `index.html` file that viewers can open in their browser. So, I think it is more likely that the user would just try to copy the QC folder (with `index.html`) from the container to the local machine using `docker cp`, then open the `index.html` file normally.

I think that you probably can go ahead and omit the instructions. Thank you for double-checking! :hearts:

```suggestion
```

joshuacwnewton commented on documentation/source/user_section/installation/windows.rst:
```diff
@@ -343,53 +327,32 @@ First, save your Docker image:
 
-      docker commit <CONTAINER_ID> <YOUR_NAME>/<DISTROS>:<VERSION>
-
-#. Install Xming
-#. Connect to it using Xming/SSH:
-
-   - If you are using Docker Desktop, please download and run (double click) the following script: :download:`sct-win.xlaunch<../../../../contrib/docker/sct-win.xlaunch>`.
-   - If you are using Docker Toolbox, please download and run the following script instead: :download:`sct-win_docker_toolbox.xlaunch<../../../../contrib/docker/sct-win_docker_toolbox.xlaunch>`
-   - If this is the first time you have done this procedure, the system will ask you if you want to add the remote PC (the docker container) as trust pc, type yes. Then type the password to enter the docker container (by default sct).
-
-**Troubleshooting:**
-
-The graphic terminal emulator LXterminal should appear (if not check the task bar at the bottom of the screen), which allows copying and pasting commands, which makes it easier for users to use it. If there are no new open windows:
-
-- Please download and run the following file: :download:`Erase_fingerprint_docker.sh<../../../../contrib/docker/Erase_fingerprint_docker.sh>`
```
If these helper scripts are no longer necessary, we may want to consider removing them from the SCT repository as well as part of this PR (assuming that they are used in no other places).

joshuacwnewton reviewed (approved):
Apart from the remaining changes, the Linux/Windows instructions LGTM, so I'm happy to give my approval.

I'll wait for approval on the Mac side of things before merging, though.

Thank you so much again for your contributions! :hearts:

mguaypaq commented on documentation/source/user_section/installation/mac.rst:
```diff
@@ -136,61 +136,43 @@ In the context of SCT, it can be used:
    # Install SCT
    git clone https://github.com/spinalcordtoolbox/spinalcordtoolbox.git sct
    cd sct
    ./install_sct -y
```
Done!

mguaypaq reviewed (approved):
Thanks so much for this documentation update, @co-decode!

I just ran the Mac instructions on one of the lab computers. Things basically worked, but I added a commit to the PR to clarify a few of the steps. I noticed a few other things, but I think they're out of scope for this PR; we should just merge it as-is, since it's a big improvement. :tada:

- For both linux and mac, the "Requirements" section at the top says to install `gcc` on the system, but this isn't needed when using the Docker method (since it gets installed inside the container, not on the host system).
- I'm not sure if it's clear that the user has to run Docker Desktop and XQuartz and `xhost + 127.0.0.1` before running the SCT container _every time_, whereas the other instructions are meant to happen _just once_.
- Even after all of that, I get the green `[OK]` checks for PyQt and matplotlib, but there's some extra noise in the output:
  ```
  Check if figure can be opened with PyQt.............QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
  libGL error: No matching fbConfigs or visuals found
  libGL error: failed to load driver: swrast
  [OK]
  ```
